posthog/api/special_migration.py

```diff
@@ -0,0 +1,100 @@
+from posthog.special_migrations.runner import MAX_CONCURRENT_SPECIAL_MIGRATIONS, is_migration_in_range
+from posthog.special_migrations.utils import force_rollback_migration, force_stop_migration, trigger_migration
+
+
+class SpecialMigrationSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = SpecialMigration
+        fields = [
+            "id",
+            "name",
+            "description",
+            "progress",
+            "status",
+            "current_operation_index",
+            "current_query_id",
+            "celery_task_id",
+            "started_at",
+            "finished_at",
+            "last_error",
+            "posthog_max_version",
+            "posthog_min_version",
+        ]
+        read_only_fields = [
+            "id",
+            "name",
+            "description",
+            "progress",
+            "status",
+            "current_operation_index",
+            "current_query_id",
+            "celery_task_id",
+            "started_at",
+            "finished_at",
+            "last_error",
+            "posthog_max_version",
+            "posthog_min_version",
+        ]
+
+
+class SpecialMigrationsViewset(StructuredViewSetMixin, viewsets.ModelViewSet):
+    queryset = SpecialMigration.objects.all()
+    permission_classes = [StaffUser]
+    serializer_class = SpecialMigrationSerializer
+
+    @action(methods=["POST"], detail=True)
+    def trigger(self, request, **kwargs):
+        if len(get_all_running_special_migrations()) >= MAX_CONCURRENT_SPECIAL_MIGRATIONS:
+            return response.Response(
+                {
+                    "success": False,
+                    "error": f"No more than {MAX_CONCURRENT_SPECIAL_MIGRATIONS} special migration can run at once.",
+                },
+                status=400,
+            )
+
+        migration_instance = self.get_object()
+
+        if not is_migration_in_range(migration_instance.posthog_min_version, migration_instance.posthog_max_version):
+            return response.Response(
+                {
+                    "success": False,
+                    "error": f"Can't run migration. Minimum PostHog version: {migration_instance.posthog_min_version}. Maximum PostHog version: {migration_instance.posthog_max_version}",
+                },
+                status=400,
+            )
+
+        migration_instance.status = MigrationStatus.Starting
+        migration_instance.save()
+
+        trigger_migration(migration_instance)
+        return response.Response({"success": True}, status=200)
+
+    # DANGEROUS! Can cause another task to be lost
+    @action(methods=["POST"], detail=True)
+    def force_stop(self, request, **kwargs):
+        migration_instance = self.get_object()
+        if migration_instance.status != MigrationStatus.Running:
+            return response.Response(
+                {"success": False, "error": "Can't stop a migration that isn't running.",}, status=400,
+            )
```

Comment:
drop this & lets try to stop it anyway since it's `force` aka regardless of everything else run it, imagine when the db state for status got messed up somehow

Reply:
🤔 good point. That seems like reasonable and expected behavior to me.

Reply:
Well... so I think your point about the rollback and the loose use of the term "force" is extremely valid.

Here is the one place I would push back against it. Conceptually, one can only stop something that has started, so it makes sense to me that you "Can't stop a migration that isn't running", even if forced. It seems to me valid that a button that "force turns off the lights" wouldn't do anything if the lights were already off.

This is also important here because force stopping means telling Celery to kill a process, which we don't want to do unless necessary, hence "force". Happy to change the name though